accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+currently quite limited.
+
+Users can enable core dump generation by passing a --coredump-on-trap flag to
+the cli, or by directly setting `coredump_on_trap` to `true` on the
+`wasmtime::Config`. Wasmtime will then generate a core dump on a trap, which
+will then enable offline debugging of the error. As other embedders of Wasmtime
+will likely need more fine-grained control over the ability to produce core
+dumps, we will also add a callback api to enable the decision to capture a core
+dump to have more context present. For example, you may want to capture exactly
+one core dump per customer and only generate a new one if there is not currently
+one cached, or would like to allow the disabling of core dump generation
+entirely for a subset of the wasm modules being run.
+
+Preserving core dump information (ie. locals and operand stack) is somewhat
+straightforward in the case of the baseline compiler, Winch, because that
+information isn’t at risk of getting optimized away. Preserving the same
+information in Cranelift is less straightforward and will require additional
+bookkeeping. The general approach we can take for that is to have the
+cranelift-wasm frontend annotate wasm locals to be preserved, and for the
+operand stack we can use our existing system of preserving metadata in egraphs.
+This should allow us to have best-effort core dumps from cranelift without
+overly pessimizing optimization.
+
+While it will be possible to generate core dumps from code compiled with
+Cranelift, we do accept that it won’t have the same level of detail available to
+Winch. As the wasm core dump format is defined in terms of the wasm execution
+model and is independent of the strategy used to execute that wasm, we will be
+able to use core dumps generated by code produced by either Cranelift or Winch
+when debugging. Being able to use a core dump generated by code compiled with
+Cranelift when debugging with Winch is extremely powerful, as it means that we
+don’t need to sacrifice the performance of generated code to enable debugging
+with a core dump. Additionally, as Winch will have more information available
+when debugging, the core dump produced by code compiled with either backend will
+be sufficient to get the debugger into a state where we have high-fidelity
+information available.
+
+## Debug Adapter Protocol
+
+We plan to add initial support for the [Debug Adapter Protocol
+(DAP)](https://microsoft.github.io/debug-adapter-protocol/) into wasmtime. This
+will involve adding an execution mode like `wasmtime --dap` or `wasmtime
+--debug`, which will expose a debug adapter. The idea here is to allow DAP
+clients to consume core dump information from Wasmtime, enabling users to start
+doing post-mortem debugging, and acting as a forcing function for us to
+implement and verify basic protocol support before moving onto live debugging
+use-cases.
+
+Why the debug adapter protocol? [Like the Language Server Protocol
```

Comment:
Gdbserver allows delegating all DWARF handling, expression parsing, conditional breakpoints, scripting, ... to GDB. The debug adapter protocol requires the DAP server (like wasmtime here) to implement this all afaik.

Reply:
That is true for DWARF and compiled-to-Wasm programs, but in order to provide debugging of interpreted-within-Wasm programs, we need to be in control of all this and can't rely on gdb to do it for us.

Reply:
Gdb and the debug adapter components would need roughly the same interface to interact with wasmtime, right? (stepping, breakpoints, inspecting wasm locals, ...) Could we export the interface imported by debug adapter components using the gdbserver interface? That way you can choose gdb for debugging with DWARF and the right debug adapter for interpreted code.

Reply:
One complication I see with that is that gdb currently doesn't have a wasm target and thus needs the dwarf translation we currently do, so either we add a wasm target to gdb or do the translation in the gdbserver adapter.